microsoft/testsuites/vm_extensions/custom_script.py

```diff
@@ -0,0 +1,533 @@
+from lisa.sut_orchestrator.azure.features import AzureExtension
+from lisa.sut_orchestrator import AZURE
+from microsoft.testsuites.vm_extensions.common import (
+    CommandInfo,
+    retrieve_storage_blob_url,
+    retrieve_storage_account_name_and_key,
+    verify_waagent_version_supported,
+)
+
+
+def _create_and_verify_extension_run(
+    node: Node,
+    settings: Optional[Dict[str, Any]] = None,
+    protected_settings: Optional[Dict[str, Any]] = None,
+    execute_commands: Optional[List[CommandInfo]] = None,
+    assert_exception: Any = None,
+) -> None:
+    extension = node.features[AzureExtension]
+
+    def enable_extension() -> Any:
+        result = extension.create_or_update(
+            name="CustomScript",
+            publisher="Microsoft.Azure.Extensions",
+            type_="CustomScript",
+            type_handler_version="2.1",
+            auto_upgrade_minor_version=True,
+            settings=settings,
+            protected_settings=protected_settings,
+        )
+        return result
+
+    if assert_exception:
+        assert_that(enable_extension).raises(assert_exception).when_called_with()
+    else:
+        result = enable_extension()
+        assert_that(result["provisioning_state"]).described_as(
+            "Expected the extension to succeed"
+        ).is_equal_to("Succeeded")
+
+    if execute_commands:
+        for command_info in execute_commands:
+            node.execute(
+                command_info.command,
+                shell=True,
+                expected_exit_code=command_info.expected_exit_code,
+                expected_exit_code_failure_message=command_info.failure_message,
+            )
+
+
+@TestSuiteMetadata(
+    area="vm_extension",
+    category="functional",
+    description="""
+    This test suite tests the functionality of the Custom Script VM extension.
+
+    File uri is a public Azure storage blob uri unless mentioned otherwise.
+    File uri points to a linux shell script unless mentioned otherwise.
+
+    It has 12 test cases to verify if CSE runs as intended when provided:
+        1. File uri and command in public settings
+        2. Two file uris and command for downloading both scripts in public settings
+        3. File uri and command in both public and protected settings (should fail)
+        4. File uri without a command or base64 script (should fail)
+        5. Both base64 script and command in public settings (should fail)
+        6. File uri and base64 script in public settings
+        7. File uri and gzip'ed base64 script in public settings
+        8. File uri and command in protected settings
+        9. Private file uri without sas token or credentials (should fail)
+        10. Private file uri with storage account credentials
+        11. Private sas file uri and command in public settings
+        12. File uri (pointing to python script) and command in public settings
+    """,
+    requirement=simple_requirement(unsupported_os=[BSD]),
```

Comment:
The requirement cannot be specified in both test suite and test case level. The test suite level is a default value, if it's not specified at test case level. The value [BSD] won't be kept, when test case level rewritten it. But if you have multiple test cases share the same requirements, it's useful to reduce repeat defining on each test case. If some test cases are different with most, you can define test case level for them.